packages/core/src/lib/utils/webauthn-utils.ts

```diff
@@ -0,0 +1,440 @@
+  }
+
+  // Verify the response
+  let verification: VerifiedRegistrationResponse
+  try {
+    const relayingParty = getRelayingParty(options)
+    verification = await verifyRegistrationResponse({
+      ...provider.verifyRegistrationOptions,
+      expectedChallenge,
+      response: data as RegistrationResponseJSON,
+      expectedOrigin: relayingParty.origin,
+      expectedRPID: relayingParty.id,
+    })
+  } catch (e: any) {
+    throw new WebAuthnVerificationError(e)
+  }
+
+  // Make sure the response was verified
+  if (!verification.verified || !verification.registrationInfo) {
+    throw new WebAuthnVerificationError("WebAuthn registration response could not be verified.")
+  }
+
+  // Build a new account
+  const account = {
+    providerAccountId: toBase64(verification.registrationInfo.credentialID),
+    provider: options.provider.id,
+    type: provider.type,
+  }
+
+  // Build a new authenticator
+  const authenticator = {
+    providerAccountId: account.providerAccountId,
+    counter: verification.registrationInfo.counter,
+    credentialID: toBase64(verification.registrationInfo.credentialID),
+    credentialPublicKey: toBase64(verification.registrationInfo.credentialPublicKey),
+    credentialBackedUp: verification.registrationInfo.credentialBackedUp,
+    credentialDeviceType: verification.registrationInfo.credentialDeviceType,
+    transports: transportsToString((data as RegistrationResponseJSON).response.transports as AuthenticatorTransport[])
+  }
+
+  // Return created stuff
+  return {
+    user,
+    account,
+    authenticator,
+  }
+}
+
+
+/**
+ * Generates WebAuthn authentication options.
+ * 
+ * @param options - The internal options for WebAuthn.
+ * @param user - Optional user information.
+ * @returns The authentication options.
+ */
+async function getAuthenticationOptions(options: InternalOptionsWebAuthn, user?: User) {
+  const { provider, adapter } = options
+
+  // Get the user's authenticators.
+  const authenticators = user && user["id"] ?
+    await adapter.listAuthenticatorsByUserId(user.id) :
+    null
+
+  const relayingParty = getRelayingParty(options)
+
+  // Return the authentication options.
+  return await generateAuthenticationOptions({
+    ...provider.authenticationOptions,
+    rpID: relayingParty.id,
+    allowCredentials: authenticators?.map((a) => ({
+      id: fromBase64(a.credentialID),
+      type: "public-key",
+      transports: stringToTransports(a.transports),
+    })),
+  })
+}
+
+
+/**
+ * Generates WebAuthn registration options.
+ * 
+ * @param options - The internal options for WebAuthn.
+ * @param user - The user information.
+ * @returns The registration options.
+ */
+async function getRegistrationOptions(options: InternalOptionsWebAuthn, user: User & { email: string }) {
+  const { provider, adapter } = options
+
+  // Get the user's authenticators.
+  const authenticators = user["id"] ? await adapter.listAuthenticatorsByUserId(user.id) : null
```

Comment:
would it make sense to throw if authenticators is an empty array or null? It means that the user attemped to authenticate without prior registering.

Reply:
I don't think so. In this case all that's gonna happen is `allowedCredentials` returns empty and the browser informs the user the server has no authenticators for them. It won't be an invalid state or anything like that, so no need to throw

Reply:
perfect!